### DISCUSSION TITLE
How do I write a dynamic file to disk that is retained after rendering?

### DISCUSSION THREAD

**Discussion by coatless** [quarto-dev/quarto-cli#5796] [Q&A]
### Description

Inside a Lua filter, I would like to generate a file that can be attached on an on-demand basis. So, I'm looking to add a dynamic "configuration" file. 

For example, I'd like the Lua filter to be able to write out a configuration file like:

```yaml
- name: config
- property:
   - render: true 
```

Is there a good way to go about this currently? I think all files largely have to be "fully" present to be included with a resource call.

**Comment by cscheid** [comment]
Can you explain in a little more detail what you're trying to achieve?

**Reply by coatless** [reply]
I'm aiming to use dynamically written configuration files to get around the raw block assertion issue that arose in #5666.  So, the workflow would simplify from using `quarto.doc.include_text()` to `quarto.doc.include_file()`. Moreover, we can hopefully just write the configuration files _once_ (detect if file is present and skip) to improve the setup and maintenance process for large book projects.

**Comment by dragonstyle** [comment]
Quarto 1.4 will include function `quarto.doc.add_resource` and `quarto.doc.add_supporting` which allow you to tell Quarto about dynamically generated files, and ensure that they are copied or moved to the same relative path in the output directory...

**Reply by coatless** [reply]
I think this is referring to #5709? 

From the NEWS entry:

> 
> - Add support `quarto.doc.add_resource` and `quarto.doc.add_supporting`. 
>   - `add_resource` will add a resouce file to the current render, copying that file to the same relative location in the output directory. 
>   - `add_supporting` will add a supporting file to the current render, moving that file file to the same relative location in the output directory.

My use case would likely be under the `add_supporting()`. So, this would mean I would use Lua's _write_ functionality to write to the `_extension` directory and, then, move the file to the render directory?

**Reply by coatless** [reply]
Is there a quarto specific way of checking to see if a document exists _before_ attempting to add the resource/support file? I'd like to avoid overwriting customizations in files if already set.

**Reply by dragonstyle** [reply]
When resource/supporting files(s) are placed after rendering, the files are moved to the output directory using the the relative path to the input file that provided the supporting / resource file(s). So I'd generally say you should write the file to wherever you'd like relative to the rendered input file (and it will be in the same position in the output directory after rendering). Does that make sense?

It looks to me like we have some internal tools for working with files in LUA that we don't currently expose in our LUA API. We could take a look at exposing those to simplify this...
